### DISCUSSION TITLE
Extend LocalizedString type to allow empty string

### DISCUSSION THREAD

**Discussion by maxpatiiuk** [ivanhofer/typesafe-i18n#676] [DX improvements]
Would it be possible to make LocalizedString typescript type that is provided by this library consider an empty string to be localized?

### Example implementation

That might look like the following:

```ts
// Currently:
export type LocalizedString = string & {
    readonly [localized]: unknown;
};
// Requested:
export type LocalizedString = 
  | ''
  | (string & {
    readonly [localized]: unknown;
  });
```

### Benefits
Empty string is often used as a placeholder/default value for a place that will store localized string.
-  Good example is React state:
    ```ts
    // Without this change we need a type assertion:
    const [userName, setUserName] = React.useState('' as LocalizedString);
    // With this change
    const [userName, setUserName] = React.useState<LocalizedString>('');
    ```
- Another use case:
   ```ts
   declare const stringOrUndefined: LocalizedStirng | undefined;
   declare const capitalize = (b: LocalizedString) => LocalizedString;   

   // Without this change we need a type assertion:
   capitalize(stringOrUndefined ?? ('' as LocalizedString));
   // With this change:
   capitalize(stringOrUndefined ?? '');
   ```

### Workarounds

So far I created a `declarations.d.ts` file like this to overwrite the type exported by this library:
```ts
import type { localized } from 'typesafe-i18n/types/runtime/src/core';

declare module 'typesafe-i18n' {
  export type LocalizedString =
    | ''
    | (string & {
        readonly [localized]: unknown;
      });
}
```

Unfortunately, not every tool supports declarations file (ts-node requires extra options, typescript-eslint has some weirdnesses; my co-workers reported problems in vs code too - the above fix doesn't work for them, even though tsc is fine with it)

### Drawbacks
There are cases where TypeScript may expand `''` to mean `string`, thus negating the benefit of `LocalizedString` type.

**Comment by ivanhofer** [comment]
Hi @maxxxxxdlp, thanks for your suggestion, but I prefer not to do so.
Types should be as strict as possible, and you can opt out any time by using a cast (`as LocalizedString`) or a utility function that does the cast (`const l = (value: string) => value as LocalizedString`).
Also the workaround you already have mentioned is a good way to customize the behavior if you really want it in your projects. It is unfortunate but not `typesafe-i18n`s fault that the tooling does not support standard `TypeScript` features.

**Reply by maxpatiiuk** [reply]
No problem, thanks!
